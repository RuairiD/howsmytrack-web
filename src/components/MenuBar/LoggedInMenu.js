import React from "react";

import { Badge, Menu, Icon } from "antd";
import { A, Span } from "lemon-reset";

import UserDetails from "./UserDetails";

const LoggedInMenu = ({ onMenuClick, username, rating, notifications }) => (
    <Menu
        theme="light"
        onClick={onMenuClick}
        style={{
            marginBottom: "auto",
            borderRight: 0,
        }}
    >
        <UserDetails
            username={username}
            rating={rating}
        />
        <Menu.Item key="newRequest">
            <Icon type="plus" />
            <Span>New Request</Span>
        </Menu.Item>
        <Menu.Item key="feedbackGroups">
            <A href="/groups">
                <Icon type="team" />
                <Span>Your Groups</Span>
                <Badge
                    count={notifications}
                    style={{ marginLeft: "0.5em" }}
                />
            </A>
        </Menu.Item>
        <Menu.Item key="faq">
            <A href="/faq">
                <Icon type="question-circle" />
                <Span>FAQ</Span>
            </A>
        </Menu.Item>
        <Menu.Item key="settings">
            <A href="/settings">
                <Icon type="setting" />
                <Span>Settings</Span>
            </A>
        </Menu.Item>
        <Menu.Item key="logout">
            <Icon type="user-delete" />
            <Span>Sign Out</Span>
        </Menu.Item>
    </Menu>
);

export default LoggedInMenu;
