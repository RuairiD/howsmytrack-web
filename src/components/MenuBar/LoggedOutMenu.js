import React from "react";

import { Menu, Icon, Typography } from "antd";
import { A, Span } from "lemon-reset";

const LoggedOutMenu = ({ onMenuClick }) => (
    <Menu
        theme="light"
        onClick={onMenuClick}
        style={{
            borderRight: 0,
        }}
    >
        <Menu.Item disabled>
            <Typography.Text>Welcome!</Typography.Text>
        </Menu.Item>
        <Menu.Item key="login">
            <Icon type="user" />
            <Span>Sign In</Span>
        </Menu.Item>
        <Menu.Item key="register">
            <Icon type="user-add" />
            <Span>Register</Span>
        </Menu.Item>
        <Menu.Item key="faq">
            <A href="/faq">
                <Icon type="question-circle" />
                <Span>FAQ</Span>
            </A>
        </Menu.Item>
    </Menu>
);

export default LoggedOutMenu;
