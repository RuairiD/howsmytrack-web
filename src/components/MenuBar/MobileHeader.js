import React from "react";

import { Badge, Typography } from "antd";
import { Div, Img, Span } from "lemon-reset";

const MobileHeader = ({ mobileMenuCollapsed, notifications }) => (
    <Div style={{ display: "flex" }}>
        <Span style={{
            marginTop: "auto",
            marginBottom: "auto",
        }}
        >
            <Typography.Text>
                how's my track?
            </Typography.Text>
            {mobileMenuCollapsed && (
                <Badge
                    count={notifications}
                    style={{
                        marginLeft: "0.5em",
                    }}
                />
            )}
        </Span>
        <Img
            alt=""
            src="/logo128.png"
            style={{
                marginLeft: "auto",
                width: "32px",
                height: "32px",
            }}
        />
    </Div>
);

export default MobileHeader;
