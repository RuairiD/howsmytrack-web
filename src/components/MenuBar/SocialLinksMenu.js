import React from "react";

import { Menu, Icon } from "antd";
import { A, Div, Span } from "lemon-reset";

const SocialLinksMenu = ({ isMobile, onMenuClick }) => {
    let style = {};
    if (!isMobile) {
        style = {
            position: "absolute",
            bottom: "0px",
        };
    }
    return (
        <Div style={style}>
            <Menu
                theme="light"
                onClick={onMenuClick}
                style={{
                    borderRight: 0,
                }}
            >
                {isMobile && <Menu.Divider />}
                <Menu.Item className="ant-menu-item">
                    <A href="https://twitter.com/howsmytrackcom" target="_blank" rel="noopener noreferrer" className="d-flex">
                        <Icon className="align-middle my-auto" type="twitter" />
                        <Span>Twitter</Span>
                    </A>
                </Menu.Item>
                <Menu.Item className="ant-menu-item">
                    <A href="https://discord.gg/6jGFP6N" target="_blank" rel="noopener noreferrer" className="d-flex">
                        <Icon className="align-middle my-auto" type="robot" />
                        <Span>Discord</Span>
                    </A>
                </Menu.Item>
            </Menu>
        </Div>
    );
};

export default SocialLinksMenu;
