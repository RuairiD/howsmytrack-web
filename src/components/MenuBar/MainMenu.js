import React from "react";

import { Div } from "lemon-reset";

import LoggedInMenu from "./LoggedInMenu";
import LoggedOutMenu from "./LoggedOutMenu";
import SocialLinksMenu from "./SocialLinksMenu";

const MainMenu = ({
    isMobile,
    onMenuClick,
    username,
    rating,
    notifications,
}) => (
    <Div>
        {username && (
            <LoggedInMenu
                onMenuClick={onMenuClick}
                username={username}
                rating={rating}
                notifications={notifications}
            />
        )}
        {!username && (
            <LoggedOutMenu
                onMenuClick={onMenuClick}
            />
        )}
        <SocialLinksMenu
            isMobile={isMobile}
            onMenuClick={onMenuClick}
        />
    </Div>
);

export default MainMenu;
