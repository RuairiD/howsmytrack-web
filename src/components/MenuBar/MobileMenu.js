import React from "react";

import { Collapse, Icon } from "antd";

import MobileHeader from "./MobileHeader";
import MainMenu from "./MainMenu";

const MobileMenu = ({
    onMenuClick,
    username,
    rating,
    notifications,
    onCollapseChange,
    mobileMenuCollapsed,
}) => (
    <Collapse
        className="mobile-menu-collapse"
        bordered={false}
        expandIcon={({ isActive }) => (
            <Icon type="menu" rotate={isActive ? 90 : 0} />
        )}
        onChange={onCollapseChange}
    >
        <Collapse.Panel
            header={(
                <MobileHeader
                    mobileMenuCollapsed={mobileMenuCollapsed}
                    notifications={notifications}
                />
            )}
            key="1"
        >
            <MainMenu
                isMobile
                onMenuClick={onMenuClick}
                username={username}
                rating={rating}
                notifications={notifications}
            />
        </Collapse.Panel>
    </Collapse>
);

export default MobileMenu;
