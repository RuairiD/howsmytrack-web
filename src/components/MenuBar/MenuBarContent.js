import React, { useState } from "react";
import ReactGA from "react-ga";
import { useQuery } from "react-query";
import axios from "axios";

import { Spin } from "antd";
import { A, Div, Img } from "lemon-reset";
import apiRoot from "../../apiRoot";
import MobileMenu from "./MobileMenu";
import MainMenu from "./MainMenu";
import FeedbackRequestModal from "../FeedbackRequestModal/FeedbackRequestModal";
import LoginModal from "../LoginModal/LoginModal";
import RegisterModal from "../RegisterModal/RegisterModal";

const USER_DETAILS_QUERY = `query UserDetails {
  userDetails {
    username
    rating
    notifications
  }
}`;

const GA_MENU_BAR_CATEGORY = "menubar";

const MenuBarContent = ({ isMobile }) => {
    const [isFeedbackRequestModalVisible, setIsFeedbackRequestModalVisible] = useState(false);
    const [isLoginModalVisible, setIsLoginModalVisible] = useState(false);
    const [isRegisterModalVisible, setIsRegisterModalVisible] = useState(false);
    const [mobileMenuCollapsed, setMobileMenuCollapsed] = useState(true);

    const logout = () => axios.get(`${apiRoot}/logout/`);

    const menuActions = {
        newRequest: () => {
            setIsFeedbackRequestModalVisible(true);
        },
        login: () => {
            setIsLoginModalVisible(true);
        },
        register: () => {
            setIsRegisterModalVisible(true);
        },
        logout: () => {
            logout().then(() => window.location.assign("/"));
        },
    };

    const { isLoading, data } = useQuery([USER_DETAILS_QUERY], () => (
        axios.post(`${apiRoot}/graphql/`, {
            query: USER_DETAILS_QUERY,
        }).then((response) => response.data.data.userDetails)
    ));

    let username = null;
    let rating = null;
    let notifications = null;
    if (!isLoading) {
        if (!data) {
            // Clear the JWT cookie in case it was mangled somehow.
            logout();
        } else {
            username = data.username;
            rating = data.rating;
            notifications = data.notifications;
            ReactGA.set({
                username: data.username,
            });
        }
    }

    const onMenuClick = (event) => {
        ReactGA.event({
            category: GA_MENU_BAR_CATEGORY,
            action: event.key,
        });
        const menuAction = menuActions[event.key];
        if (menuAction) {
            menuAction();
        }
    };

    const onFeedbackRequestModalCancel = () => {
        setIsFeedbackRequestModalVisible(false);
    };

    const onLoginModalCancel = () => {
        setIsLoginModalVisible(false);
    };

    const onRegisterModalCancel = () => {
        setIsRegisterModalVisible(false);
    };

    const onCollapseChange = (key) => {
        setMobileMenuCollapsed(key.length === 0);
    };

    return (
        <Spin spinning={isLoading}>
            <Div className="menu-bar">
                {!isMobile && (
                    <A href="/" className="full-logo">
                        <Img alt="how's my track" src="/logo512.png" width="200px" style={{ padding: "0.5em", marginLeft: "auto", marginRight: "auto" }} />
                    </A>
                )}
                {isMobile && (
                    <MobileMenu
                        onMenuClick={onMenuClick}
                        username={username}
                        rating={rating}
                        notifications={notifications}
                        onCollapseChange={onCollapseChange}
                        mobileMenuCollapsed={mobileMenuCollapsed}
                    />
                )}
                {!isMobile && (
                    <MainMenu
                        onMenuClick={onMenuClick}
                        username={username}
                        rating={rating}
                        notifications={notifications}
                    />
                )}

                <FeedbackRequestModal onCancel={onFeedbackRequestModalCancel} isVisible={isFeedbackRequestModalVisible} />
                <LoginModal onCancel={onLoginModalCancel} isVisible={isLoginModalVisible} />
                <RegisterModal onCancel={onRegisterModalCancel} isVisible={isRegisterModalVisible} />
            </Div>
        </Spin>
    );
};

export default MenuBarContent;
