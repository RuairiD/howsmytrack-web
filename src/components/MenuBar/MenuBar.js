import React from "react";

import { Affix } from "antd";
import { Div } from "lemon-reset";
import MenuBarContent from "./MenuBarContent";

type Props = {
    isMobile: boolean,
};

const MenuBar = ({ isMobile }: Props) => {
    /*
     * Component for displaying page sidebar with menu links, or for mobile,
     * displaying a menu at the top of the screen.
     */

    // Wrapping this in Div prevents forwardRef related errors being raised by
    // ant-design; an inelegant solution but an unobtrusive one as well.
    const menuBarContent = (
        <Div>
            <MenuBarContent
                isMobile={isMobile}
            />
        </Div>
    );

    if (isMobile) {
        return menuBarContent;
    }
    return (
        <Affix>
            {menuBarContent}
        </Affix>
    );
};

export default MenuBar;
