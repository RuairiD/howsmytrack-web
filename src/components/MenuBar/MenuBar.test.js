import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import MenuBar from "./MenuBar";

describe("MenuBar", () => {
    it("renders MenuBarContent within an Affix for non-mobile clients", () => {
        const wrapper = shallow(
            <MenuBar />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders solo MenuBarContent mobile clients", () => {
        const wrapper = shallow(
            <MenuBar isMobile />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
