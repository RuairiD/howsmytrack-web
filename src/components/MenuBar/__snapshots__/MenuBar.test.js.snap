// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuBar renders MenuBarContent within an Affix for non-mobile clients 1`] = `
<Affix
  target={[Function]}
>
  <Div
    className=""
  >
    <MenuBarContent />
  </Div>
</Affix>
`;

exports[`MenuBar renders solo MenuBarContent mobile clients 1`] = `
<Div
  className=""
>
  <MenuBarContent
    isMobile={true}
  />
</Div>
`;
